### PR TITLE
[rllib] Behavior Cloning

### PR DESCRIPTION
**Pull request by royf**
Implements the Behavior Cloning (BC) algorithm in RLlib.

To run it:
1. Train a supervisor with any algorithm.
2. Rollout the trained supervisor to demonstrate the behavior to clone:
`rollout.py <checkpoint_path> --run <training_algorithm> --env <env_name> --steps <demonstration_steps> --out <rollouts_path>`
3. Train from these demonstrations:
`train.py --run BC --env <env_name> --config '{"dataset_path": <rollouts_path>}'`

Tested on FrozenLake-v0 with 100K demonstration steps of a DQN supervisor.

### CODE REVIEW

**Comment by richardliaw** [comment]
ok to test

File: python/ray/rllib/a3c/a3c.py [hunk 1]
```diff
@@ -66,6 +66,7 @@
 class A3CAgent(Agent):
     _agent_name = "A3C"
     _default_config = DEFAULT_CONFIG
+    _allow_unknown_configs = True
```
Comment (richardliaw) [inline python/ray/rllib/a3c/a3c.py]: probably want to keep this as `False`

File: python/ray/rllib/rollout.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
Comment (richardliaw) [inline python/ray/rllib/rollout.py]: I believe this is very similar to `eval.py` - perhaps just modify that instead of introducing a new file?

File: python/ray/rllib/optimizers/optimizer.py [hunk 3]
```diff
@@ -7,8 +7,8 @@ class Optimizer(object):
     """RLlib optimizers encapsulate distributed RL optimization strategies.
 
     For example, AsyncOptimizer is used for A3C, and LocalMultiGPUOptimizer is
-    used for PPO. These optimizers are all pluggable however, it is possible
-    to mix as match as needed.
+    used for PPO. These optimizers are all pluggable and it is possible
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/optimizer.py]: "all pluggable, and"

File: python/ray/rllib/bc/tfpolicy.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            bs = tf.to_float(tf.shape(self.x)[0])
+            tf.summary.scalar("model/policy_loss", self.pi_loss / bs)
+            tf.summary.scalar("model/grad_gnorm", tf.global_norm(self.grads))
+            tf.summary.scalar("model/var_gnorm", tf.global_norm(self.var_list))
+            self.summary_op = tf.summary.merge_all()
+
+        # TODO(rliaw): Can consider exposing these parameters
+        self.sess = tf.Session(graph=self.g, config=tf.ConfigProto(
+            intra_op_parallelism_threads=1, inter_op_parallelism_threads=2,
+            gpu_options=tf.GPUOptions(allow_growth=True)))
```
Comment (richardliaw) [inline python/ray/rllib/bc/tfpolicy.py]: is this the only change in this file compared to `a3c/tfpolicy`? If so, we should just expose these parameters via the default config in A3C and avoid repeating too much code.

File: python/ray/rllib/bc/policy.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from ray.rllib.bc.tfpolicy import TFPolicy
+from ray.rllib.models.catalog import ModelCatalog
+
+
+class Policy(TFPolicy):
+    def __init__(self, registry, ob_space, ac_space, config, **kwargs):
```
Comment (richardliaw) [inline python/ray/rllib/bc/policy.py]: You don't need to include this (I'm pretty sure)

File: python/ray/rllib/bc/policy.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+from __future__ import absolute_import
```
Comment (richardliaw) [inline python/ray/rllib/bc/policy.py]: In the spirit of reducing repetition, this file could probably be extended from `a3c/shared_model.py` with a modification of the `setup_loss`

File: python/ray/rllib/a3c/tfpolicy.py [hunk 7]
```diff
@@ -28,7 +28,7 @@ def __init__(self, registry, ob_space, action_space, config,
             self.setup_gradients()
             self.initialize()
 
-    def _setup_graph(self):
+    def _setup_graph(self, ob_space, ac_space):
```
Comment (richardliaw) [inline python/ray/rllib/a3c/tfpolicy.py]: This is good, thanks.

File: python/ray/rllib/bc/bc.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        self.optimizer.step()
+        metric_lists = [re.get_metrics.remote() for re in self.remote_evaluators]
+        total_samples = 0
+        total_loss = 0
+        for metrics in metric_lists:
+            for m in ray.get(metrics):
+                total_samples += m["num_samples"]
+                total_loss += m["loss"]
+        result = TrainingResult(
+            episode_reward_mean=total_loss / total_samples,
```
Comment (richardliaw) [inline python/ray/rllib/bc/bc.py]: I believe there's a specific field for `loss` separate from `reward` - check `tune/result.py`

File: python/ray/rllib/bc/bc.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        self.optimizer = AsyncOptimizer(
+            self.config["optimizer"], self.local_evaluator,
+            self.remote_evaluators)
+
+    def _train(self):
+        self.optimizer.step()
+        metric_lists = [re.get_metrics.remote() for re in self.remote_evaluators]
+        total_samples = 0
+        total_loss = 0
+        for metrics in metric_lists:
```
Comment (richardliaw) [inline python/ray/rllib/bc/bc.py]: probably just need to do `for m in ray.get(metric_lists)`

File: python/ray/rllib/bc/experience_dataset.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+from __future__ import absolute_import
```
Comment (richardliaw) [inline python/ray/rllib/bc/experience_dataset.py]: We will need to extend the functionality of this class to work on multiple machines.

**Review by richardliaw — CHANGES_REQUESTED** [review]
Overall this is neat! Thanks for contributing this. Couple comments; would be great if we reduced redundancy.

File: python/ray/rllib/a3c/a3c.py (hunk 1 above)
Comment (royf) [reply python/ray/rllib/a3c/a3c.py]: without this I got `Unknown config parameter vf_loss_coeff`

File: python/ray/rllib/bc/bc.py (hunk 9 above)
Comment (royf) [reply python/ray/rllib/bc/bc.py]: `metric_lists` is a list of lists (one per remove evaluator) of dicts (one per finished batch); this parallels `A3CAgent._fetch_metrics_from_remote_evaluators`

File: python/ray/rllib/bc/experience_dataset.py (hunk 10 above)
Comment (royf) [reply python/ray/rllib/bc/experience_dataset.py]: yea, it's currently loading the dataset from a specified path; we may want to ray.put the dataset, would that be a problem with really big datasets (say up to several 10GB)? what's the alternative?

File: python/ray/rllib/bc/policy.py (hunk 6 above)
Comment (royf) [reply python/ray/rllib/bc/policy.py]: indeed, they should both be external to any specific algo; this is one of the issues I wanted to open

File: python/ray/rllib/bc/tfpolicy.py (hunk 4 above)
Comment (royf) [reply python/ray/rllib/bc/tfpolicy.py]: the bigger change is removing the `assert` in `__init__`; and I didn't want to create a dependency of BC on A3C; we should refactor the policy to be external to either of these

File: python/ray/rllib/rollout.py (hunk 2 above)
Comment (royf) [reply python/ray/rllib/rollout.py]: that's reasonable; `eval.py` is missing `ModelCatalog.get_preprocessor_as_wrapper` though, right?

File: python/ray/rllib/a3c/a3c.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/rllib/a3c/a3c.py]: really? `vf_loss_coeff` is defined above in line 40; can you help me reproduce this error?

File: python/ray/rllib/bc/bc.py (hunk 9 above)
Comment (richardliaw) [reply python/ray/rllib/bc/bc.py]: I see

File: python/ray/rllib/bc/experience_dataset.py (hunk 10 above)
Comment (richardliaw) [reply python/ray/rllib/bc/experience_dataset.py]: depends on how much memory you allocate for your instance, but you could always partition the dataset so that the object store acts more as a buffer. I think `ray.put` should be fine.

File: python/ray/rllib/bc/policy.py (hunk 6 above)
Comment (richardliaw) [reply python/ray/rllib/bc/policy.py]: yeah, the model/framework-dependent code hasn't been refactored yet.

File: python/ray/rllib/rollout.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/rllib/rollout.py]: yeah, the wrapper would be a nice addition

File: python/ray/rllib/bc/tfpolicy.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/rllib/bc/tfpolicy.py]: I see; there is still a dependency on A3C via `a3c/policy.py`. I guess it would be reasonable to duplicate all of the policy code for now, @ericl?

**Review by richardliaw — COMMENTED** [review]
Responses and a few more comments. Also, it would be nice to be able to support the gym suite. Can you make sure this works for CartPole-v0, Mujoco tasks, and Atari (Pong and FrozenLake)?

Also, a test in `test/jenkins_tests/run_multi_node_tests.sh`  somehow would be great.

File: python/ray/rllib/bc/experience_dataset.py (hunk 10 above)
Comment (ericl) [reply python/ray/rllib/bc/experience_dataset.py]: Put could be hazardous if the dataset gets evicted at some point. Maybe for now just document that the files should be available on each machine.

File: python/ray/rllib/bc/tfpolicy.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/bc/tfpolicy.py]: Duplicating sounds good, there's not a lot.

Btw, shouldn't we combine Policy and TFPolicy into one class? There doesn't seem to be much value in separating them.

**Comment by ericl** [comment]
Looks good. Should we add a paragraph in `doc/source/rllib.rst` on how to use BC?

File: python/ray/rllib/bc/tfpolicy.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/rllib/bc/tfpolicy.py]: Yeah, probably should combine `bc/policy.py` and `bc/tfpolicy.py`. We can leave the `a3c/policy.py` dependency as it's just an interface.

File: python/ray/rllib/a3c/a3c.py (hunk 1 above)
Comment (royf) [reply python/ray/rllib/a3c/a3c.py]: Nevermind, I guess I mixed incompatible installation and `train.py` revisions.

**Comment by royf** [comment]
Commit addresses remaining comments, except [testing](https://github.com/ray-project/ray/pull/1400#pullrequestreview-87093219) and [documentation](https://github.com/ray-project/ray/pull/1400#issuecomment-355854623).

**Review by ericl — APPROVED** [review]
This LGTM, looking forwards to using it

**Review by richardliaw — APPROVED** [review]
